pkg/info/info.go

```diff
@@ -26,53 +26,99 @@ type InfoServer struct {
 
 func NewInfoServer(
 	component string,
-	options ...func(*InfoServer),
+	options ...ServerOption,
 ) *InfoServer {
 	srv := &InfoServer{
 		component: component,
 	}
 
 	for _, o := range options {
-		o(srv)
+		o.applyServerOption(srv)
 	}
 
 	return srv
 }
 
-func WithLabelSet(getLabelSet func() []labelpb.ZLabelSet) func(*InfoServer) {
-	return func(s *InfoServer) {
-		s.getLabelSet = getLabelSet
+type ServerOption interface {
+	applyServerOption(*InfoServer)
+}
+
+type serverOptionFunc func(*InfoServer)
+
+func (fn serverOptionFunc) applyServerOption(cfg *InfoServer) {
+	fn(cfg)
+}
+
+func WithLabelSetFunc(getLabelSet ...func() []labelpb.ZLabelSet) ServerOption {
+	if len(getLabelSet) == 0 {
+		return serverOptionFunc(func(s *InfoServer) {
+			s.getLabelSet = func() []labelpb.ZLabelSet { return []labelpb.ZLabelSet{} }
+		})
```

Comment:
I am wondering if converting it to an `interface` is required? We can probably make it simpler by just using a type alias.

```suggestion
type ServerOptionFunc func(*InfoServer)

func WithLabelSetFunc(getLabelSet ...func() []labelpb.ZLabelSet) ServerOptionFunc {
	if len(getLabelSet) == 0 {
		return func(s *InfoServer) {
			s.getLabelSet = func() []labelpb.ZLabelSet { return []labelpb.ZLabelSet{} }
		}
```

And modifying the `NewInfoServer` just a bit.
```go
func NewInfoServer(
	component string,
	options ...ServerOptionFunc,
) *InfoServer {
	srv := &InfoServer{
		component: component,
	}

	for _, o := range options {
		o(srv)
	}

	return srv
}
```

WDYT?

Reply:
You're right, the intent was to make this so that it cannot be implemented by others, but it does not really have any value in this context.